api/kogito-services/src/main/java/org/kie/kogito/services/context/ProcessInstanceContext.java

```diff
@@ -0,0 +1,214 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.kie.kogito.services.context;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+/**
+ * Utility class for managing process instance context in logging operations.
+ * This class uses SLF4J's Mapped Diagnostic Context (MDC) to ensure process instance IDs
+ * are automatically included in log messages.
+ *
+ * When no process instance is available, an empty string is used as the default value
+ * to provide cleaner formatting and easier searching in log aggregation systems.
+ *
+ * Thread Safety: This class is thread-safe and properly manages context isolation
+ * between different threads using MDC's inherent ThreadLocal-based storage.
+ */
+public final class ProcessInstanceContext {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProcessInstanceContext.class);
+
+    private static final Map<String, ContextExtension> EXTENSIONS = new ConcurrentHashMap<>();
+
+    /**
+     * MDC key used to store the process instance ID.
+     * This key should be referenced in logging configurations.
+     */
+    public static final String MDC_PROCESS_INSTANCE_KEY = "processInstanceId";
+
+    /**
+     * Default value used when no process instance ID is available.
+     * Empty string for cleaner log formatting and easier searching.
+     */
+    public static final String GENERAL_CONTEXT = "";
+
+    // Private constructor to prevent instantiation
+    private ProcessInstanceContext() {
+        // Utility class
+    }
+
+    /**
+     * Sets the process instance ID for the current thread context.
+     * This method updates SLF4J MDC and only performs the update if the value changes.
+     *
+     * @param processInstanceId the process instance ID to set, or null to use general context
+     */
+    public static void setProcessInstanceId(String processInstanceId) {
+        String effectiveId = processInstanceId != null ? processInstanceId : GENERAL_CONTEXT;
+
+        // Only update MDC if the value is changing (optimization)
+        String currentId = MDC.get(MDC_PROCESS_INSTANCE_KEY);
+        if (!effectiveId.equals(currentId)) {
+            MDC.put(MDC_PROCESS_INSTANCE_KEY, effectiveId);
+        }
+    }
+
+    /**
+     * Gets the current process instance ID from MDC.
+     *
+     * @return the current process instance ID, or empty string if no context is set
+     */
+    public static String getProcessInstanceId() {
+        String id = MDC.get(MDC_PROCESS_INSTANCE_KEY);
+        return id != null ? id : GENERAL_CONTEXT;
+    }
+
+    /**
+     * Clears the process instance context for the current thread.
+     * This resets the MDC to the general context (empty string).
+     */
+    public static void clear() {
+        MDC.put(MDC_PROCESS_INSTANCE_KEY, GENERAL_CONTEXT);
```

Comment:
Why reset to empty string instead of removing the key from the context:

```suggestion
        MDC.remove(MDC_PROCESS_INSTANCE_KEY);
```

I think it's a better pattern to reset MDC.